### PR TITLE
refactor: Changed CTC decoder from keras backend to tf.nn

### PR DESCRIPTION
**Pull request by charlesmindee**
This PR changes CTC decoder of keras decoder which seems to be deprecated to a tf.nn ctc decoder

### CODE REVIEW

File: doctr/models/recognition/postprocessor.py [hunk 1]
```diff
@@ -52,21 +52,18 @@ def ctc_decoder(
 
         """
         batch_len = tf.cast(tf.shape(logits)[0], dtype=tf.int64)
-        sequence_len = tf.cast(tf.shape(logits)[1], dtype=tf.int64)
+        sequence_len = tf.cast(tf.shape(logits)[1], dtype=tf.int32)
 
         # computing prediction with ctc decoder
-        decoded_logits = K.ctc_decode(
-            tf.nn.softmax(logits), sequence_len * tf.ones(shape=(batch_len,), dtype=tf.int64), greedy=True
+        _prediction = tf.nn.ctc_greedy_decoder(
```
Comment (fg-mindee) [inline doctr/models/recognition/postprocessor.py:58]: if this backward compatible with both TF2.3 and TF2.4?

**Review by fg-mindee — COMMENTED** [review]
Thanks for the PR! Looks good to me, I only have a question about backward compatibility

File: doctr/models/recognition/postprocessor.py (hunk 1 above)
Comment (charlesmindee) [reply doctr/models/recognition/postprocessor.py:43]: Yes, I ran it on 2.3.1 on my computer

File: doctr/models/recognition/postprocessor.py [hunk 2]
```diff
@@ -37,21 +37,18 @@ def ctc_decoder(
...
+            tf.nn.softmax(tf.transpose(logits, perm=[1, 0, 2])),
+            sequence_len * tf.ones(shape=(batch_len,), dtype="int32"),
+            merge_repeated=True
         )
-        _predictions = tf.squeeze(decoded_logits[0])
+        _prediction = _prediction[0][0]
+        prediction = tf.sparse.to_dense(_prediction, default_value=self.num_classes)
```
Comment (fg-mindee) [inline doctr/models/recognition/postprocessor.py:49]: CI seems to fail here, I think it's `len(self.vocab) + 1`

**Review by fg-mindee — COMMENTED** [review]
Perhaps we need to resolve the alphabet issue with <eos> first. But we need to change the reference to self.num_classes

File: doctr/models/recognition/postprocessor.py [hunk 3]
```diff
@@ -37,21 +37,18 @@ def ctc_decoder(
...
         batch_len = tf.cast(tf.shape(logits)[0], dtype=tf.int64)
-        sequence_len = tf.cast(tf.shape(logits)[1], dtype=tf.int64)
+        sequence_len = tf.cast(tf.shape(logits)[1], dtype=tf.int32)
 
         # computing prediction with ctc decoder
-        decoded_logits = K.ctc_decode(
-            tf.nn.softmax(logits), sequence_len * tf.ones(shape=(batch_len,), dtype=tf.int64), greedy=True
+        _prediction = tf.nn.ctc_greedy_decoder(
+            tf.nn.softmax(tf.transpose(logits, perm=[1, 0, 2])),
+            sequence_len * tf.ones(shape=(batch_len,), dtype="int32"),
```
Comment (fg-mindee) [inline doctr/models/recognition/postprocessor.py:45]: `tf.fill(tf.shape(logits)[0], tf.shape(logits)[1])`

File: doctr/models/recognition/postprocessor.py [hunk 4]
```diff
@@ -37,21 +37,18 @@ def ctc_decoder(
 
         """
         batch_len = tf.cast(tf.shape(logits)[0], dtype=tf.int64)
```
Comment (fg-mindee) [inline doctr/models/recognition/postprocessor.py:39]: You can remove this

File: doctr/models/recognition/postprocessor.py [hunk 5]
```diff
@@ -37,21 +37,18 @@ def ctc_decoder(
...
-        sequence_len = tf.cast(tf.shape(logits)[1], dtype=tf.int64)
+        sequence_len = tf.cast(tf.shape(logits)[1], dtype=tf.int32)
```
Comment (fg-mindee) [inline doctr/models/recognition/postprocessor.py:40]: This can be removed

File: doctr/models/recognition/postprocessor.py [hunk 6]
```diff
@@ -37,21 +37,18 @@ def ctc_decoder(
...
         # computing prediction with ctc decoder
-        decoded_logits = K.ctc_decode(
-            tf.nn.softmax(logits), sequence_len * tf.ones(shape=(batch_len,), dtype=tf.int64), greedy=True
+        _prediction = tf.nn.ctc_greedy_decoder(
+            tf.nn.softmax(tf.transpose(logits, perm=[1, 0, 2])),
+            sequence_len * tf.ones(shape=(batch_len,), dtype="int32"),
+            merge_repeated=True
         )
-        _predictions = tf.squeeze(decoded_logits[0])
+        _prediction = _prediction[0][0]
```
Comment (fg-mindee) [inline doctr/models/recognition/postprocessor.py:48]: You can remove this

File: doctr/models/recognition/postprocessor.py [hunk 7]
```diff
@@ -37,21 +37,18 @@ def ctc_decoder(
...
+        sequence_len = tf.cast(tf.shape(logits)[1], dtype=tf.int32)
 
         # computing prediction with ctc decoder
-        decoded_logits = K.ctc_decode(
-            tf.nn.softmax(logits), sequence_len * tf.ones(shape=(batch_len,), dtype=tf.int64), greedy=True
+        _prediction = tf.nn.ctc_greedy_decoder(
+            tf.nn.softmax(tf.transpose(logits, perm=[1, 0, 2])),
+            sequence_len * tf.ones(shape=(batch_len,), dtype="int32"),
+            merge_repeated=True
         )
```
Comment (fg-mindee) [inline doctr/models/recognition/postprocessor.py:47]: `)[0][0]`

**Review by fg-mindee — COMMENTED** [review]
I added some other comments to improve memory efficiency

File: doctr/models/recognition/postprocessor.py (hunk 2 above)
Comment (charlesmindee) [reply doctr/models/recognition/postprocessor.py:49]: len(self.vocab) = num_classes so I think len(self.vocab)

File: doctr/models/recognition/postprocessor.py [hunk 8]
```diff
@@ -34,19 +34,14 @@ def ctc_decoder(
...
         _prediction = tf.nn.ctc_greedy_decoder(
             tf.nn.softmax(tf.transpose(logits, perm=[1, 0, 2])),
-            sequence_len * tf.ones(shape=(batch_len,), dtype="int32"),
+            tf.fill(tf.shape(logits)[0], tf.shape(logits)[1]),
             merge_repeated=True
-        )
-        _prediction = _prediction[0][0]
-        prediction = tf.sparse.to_dense(_prediction, default_value=len(self.vocab)- 1)
+        )[0][0]
+        prediction = tf.sparse.to_dense(_prediction, default_value=len(self.vocab))
```
Comment (fg-mindee) [inline doctr/models/recognition/postprocessor.py:44]: I think we are mixing PR here: for now the default_value should be kept as `len(self.vocab) - 1` ?

Comment (charlesmindee) [reply doctr/models/recognition/postprocessor.py:44]: ok
